workbench-core/authorization/src/dynamicAuthorization/ddbDynamicAuthorizationPermissionsPlugin.test.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  DynamoDBClient,
+  ServiceInputTypes,
+  ServiceOutputTypes,
+  TransactWriteItemsCommand,
+  TransactionCanceledException
+} from '@aws-sdk/client-dynamodb';
+import { AwsService } from '@aws/workbench-core-base';
+import { mockClient, AwsStub } from 'aws-sdk-client-mock';
+import { Action } from '../action';
+import { IdentityPermissionCreationError } from '../errors/identityPermissionCreationError';
+import { ThroughputExceededError } from '../errors/throughputExceededError';
+import { Effect } from '../permission';
+import { DDBDynamicAuthorizationPermissionsPlugin } from './ddbDynamicAuthorizationPermissionsPlugin';
+import { IdentityPermission, IdentityType } from './dynamicAuthorizationInputs/identityPermission';
+
+describe('DDB Dynamic Authorization Permissions Plugin tests', () => {
+  const region = 'us-east-1';
+  const ddbTableName = 'PermissionsTable';
+  const mockAuthenticatedUser = {
+    id: 'sampleUserId',
+    roles: []
+  };
+
+  let dynamoDBDynamicPermissionsPlugin: DDBDynamicAuthorizationPermissionsPlugin;
+  let awsService: AwsService;
+  let mockDDB: AwsStub<ServiceInputTypes, ServiceOutputTypes>;
+
+  const sampleGroupId = 'sampleGroup';
+  const sampleGroupType: IdentityType = 'GROUP';
+
+  const sampleAction: Action = 'CREATE';
+  const sampleEffect: Effect = 'ALLOW';
+  const sampleSubjectType = 'sampleSubjectType';
+  const sampleSubjectId = 'sampleSubjectId';
+  const sampleConditions = {};
+  const sampleFields: string[] = [];
+  const sampleDescription: string = 'sampleDescription';
+
+  const mockIdentityPermission: IdentityPermission = {
+    action: sampleAction,
+    effect: sampleEffect,
+    subjectType: sampleSubjectType,
+    subjectId: sampleSubjectId,
+    identityId: sampleGroupId,
+    identityType: sampleGroupType,
+    conditions: sampleConditions,
+    fields: sampleFields,
+    description: sampleDescription
+  } as const;
+  beforeEach(() => {
+    jest.resetModules(); // Most important - it clears the cache
+    awsService = new AwsService({ region, ddbTableName });
+    dynamoDBDynamicPermissionsPlugin = new DDBDynamicAuthorizationPermissionsPlugin({
+      awsService
+    });
+
+    mockDDB = mockClient(DynamoDBClient);
```

Comment:
Unless we're directly creating dynamo queries in AuthZ, I'd recommend mocking out the `DynamoDBService` instead of dynamo commands. By mocking something that is a dependency of that service, we're tightly coupling our implementation to the tests. Mocking the service itself means these tests are robust to any changes in how we use dynamo (or any other database that backs things in the future).

Reply:
given this ddb is the main driver for this plugin, we want to ensure certain actions are taken. such as transactWriteItem. So we are expecting these functions to be called within the ddb service

Reply:
Agreed! Those tests would be on the `dynamoDBService` though, since its a responsibility of that object. The responsibility of the `ddbDynamicAuthorizationPermissionsPlugin` is to call it's database dependency to perform db operations, not to call the `DynamoDBClient`. If we needed to change something in the future that resulted in a different type of dynamo write operation, a test-time database that would allow us to do isolated reads/writes during integration tests, or something else where the test-time and run-time dependencies are swapped, then any tests that mocks a peripheral dependency will fail even though the code may still work in prod.